src/rt/cover.c

```diff
@@ -368,323 +504,571 @@ static cover_file_t *cover_file(const loc_t *loc)
+            if (tag->data & 0x2) {
+               (ctx->flat_stats.hit_toggles)++;
+               (ctx->nested_stats.hit_toggles)++;
+               cover_append_to_chain(&(ctx->ch_toggle), true, tag,
+                                     line, COV_FLAG_TOGGLE_TO_0);
+            } else {
+               cover_append_to_chain(&(ctx->ch_toggle), false, tag,
+                                     line, COV_FLAG_TOGGLE_TO_0);
+            }
+            break;
+
+         default:
+            fatal("Unsupported type of code coverage:%d !", tag->kind);
+         }
+      }
    }
-}
 
-static void cover_index(ident_t name, const char *dir)
-{
-   char *buf = xasprintf("%s/index.html", dir);
-   FILE *fp = lib_fopen(lib_work(), buf, "w");
-   if (fp == NULL)
-      fatal("failed to create %s", buf);
-   free(buf);
-
-   cover_html_header(fp, "Coverage report for %s", istr(name));
-
-   fprintf(fp, "<h1>Coverage report for %s</h1>\n", istr(name));
-   fprintf(fp, "<div class=\"help\"><p>Select a file from the sidebar to see "
-           "annotated statement and condition coverage.</p></div>\n");
-
-   fprintf(fp, "<h2>Overall Statistics</h2>\n");
-   fprintf(fp, "<table class=\"stats\">\n");
-   fprintf(fp, "<tr><th>Metric</th><th>Covered</th><th>Total</th>"
-           "<th>Percentage</th></tr>\n");
-   cover_stat_line(fp, "Statements executed",
-                   stats.hit_stmts, stats.total_stmts);
-   cover_stat_line(fp, "Branches observed taken and not taken",
-                   stats.hit_branches, stats.total_branches);
-   cover_stat_line(fp, "Conditions evaluated to both TRUE and FALSE",
-                   stats.hit_conds, stats.total_conds);
-   fprintf(fp, "</table>\n");
-
-   fprintf(fp, "<div class=\"advert\"><p>Generated by %s\n"
-           "<br/><a href=\""PACKAGE_URL"\">"
-           PACKAGE_URL"</a></p></div>\n",
-           PACKAGE_STRING);
-
-   cover_html_footer(fp);
-
-   fclose(fp);
-}
+   cover_print_hierarchy_footer(f);
 
-void cover_report(tree_t top, cover_tagging_t *tagging,
-                  const int32_t *stmts, const int32_t *conds)
-{
-   for (unsigned i = 0; i < tagging->tags.count; i++)
-      cover_process_tag(&(tagging->tags.items[i]), stmts, conds);
+   fprintf(f, "  <h3 style=\"margin-left: " MARGIN_LEFT "px;\"> Current Instance: </h3>\n");
+   cover_print_hierarchy_header(f);
+   cover_print_hierarchy_summary(f, &(ctx->flat_stats), tag->hier, false);
+   cover_print_hierarchy_footer(f);
+
+   fprintf(f, "  <h3 style=\"margin-left: " MARGIN_LEFT "px;\"> Details: </h3>\n");
+   cover_print_hierarchy_guts(f, ctx);
+   cover_print_timestamp(f);
 
-   ident_t name = ident_strip(tree_ident(top), ident_new(".elab"));
+   fclose(f);
+   return tag;
+}
 
-   char *dir LOCAL = xasprintf("%s.cover", istr(name));
 
-   lib_t work = lib_work();
-   lib_mkdir(work, dir);
+void cover_report(const char *path, cover_tagging_t *tagging)
+{
+   char *subdir = xasprintf("%s/hier", path);
```

Comment:
`subdir` leaks I think. Put `LOCAL` after the variable name and it will be freed when it goes out of scope.

Reply:
True, added the flag. Btw. does that work accross compilers or is it GCC specific (Google says its GCC only). That brings me to a question: Why are you using GCC to build NVC while you use LLVM as code generation engine? Wouldn't it be better to use just single compiler?